### PR TITLE
refactor: em and connection generic types

### PR DESCRIPTION
**Pull request by pepakriz**

### CODE REVIEW

File: lib/EntityManager.ts [hunk 1]
```diff
@@ -5,28 +5,28 @@ import { AbstractSqlDriver, IDatabaseDriver } from './drivers';
...
-  getDriver<D extends IDatabaseDriver = IDatabaseDriver>(): D {
-    return this.driver as D;
+  getDriver(): D {
+    return this.driver;
   }
 
-  getConnection<C extends Connection = Connection>(type?: 'read' | 'write'): C {
-    return this.driver.getConnection(type) as C;
+  getConnection(type?: 'read' | 'write'): ReturnType<D['getConnection']> {
+    return this.driver.getConnection(type) as any;
```
Comment (pepakriz) [inline lib/EntityManager.ts]: Probably a bug in TS. I'll check it again later. But outer usage works correctly.

File: lib/EntityManager.ts [hunk 2]
```diff
@@ -48,7 +48,7 @@ export class EntityManager {
 
   createQueryBuilder(entityName: EntityName<AnyEntity>, alias?: string, type?: 'read' | 'write'): QueryBuilder {
     entityName = Utils.className(entityName);
-    return new QueryBuilder(entityName, this.metadata, this.driver as AbstractSqlDriver, this.transactionContext, alias, type);
+    return new QueryBuilder(entityName, this.metadata, this.driver as any as AbstractSqlDriver, this.transactionContext, alias, type);
```
Comment (pepakriz) [inline lib/EntityManager.ts]: I'll try to find a better solution.

File: lib/MikroORM.ts [hunk 3]
```diff
@@ -66,11 +66,11 @@ export class MikroORM {
   }
 
   getSchemaGenerator(): SchemaGenerator {
-    return new SchemaGenerator(this.driver as AbstractSqlDriver, this.metadata);
+    return new SchemaGenerator(this.driver as any as AbstractSqlDriver, this.metadata);
```
Comment (pepakriz) [inline lib/MikroORM.ts]: I'll try to find a better solution.

File: lib/MikroORM.ts [hunk 4]
```diff
@@ -66,11 +66,11 @@ export class MikroORM {
...
   }
 
   getEntityGenerator(): EntityGenerator {
-    return new EntityGenerator(this.driver as AbstractSqlDriver, this.config);
+    return new EntityGenerator(this.driver as any as AbstractSqlDriver, this.config);
```
Comment (pepakriz) [inline lib/MikroORM.ts]: I'll try to find a better solution.

File: lib/utils/Configuration.ts [hunk 5]
```diff
@@ -163,13 +163,17 @@ export class Configuration {
...
 
-  private initDriver(): IDatabaseDriver {
+  private initDriver(): D {
     if (!this.options.driver) {
       const driver = Configuration.PLATFORMS[this.options.type];
       this.options.driver = require('../drivers/' + driver)[driver];
     }
 
-    return new this.options.driver!(this);
+    if (!this.options.driver) {
```
Comment (pepakriz) [inline lib/utils/Configuration.ts]: Because TS don't understand of line 169

Comment (pepakriz) [reply lib/utils/Configuration.ts]: Or we can preserve the original solution.

Comment (pepakriz) [reply lib/utils/Configuration.ts]: reverted

File: lib/EntityManager.ts [hunk 6]
```diff
@@ -5,28 +5,29 @@ import { AbstractSqlDriver, IDatabaseDriver } from './drivers';
...
+import { Transaction } from './connections';
+import { MySqlDriver } from './drivers/MySqlDriver';
 
-export class EntityManager {
+export class EntityManager<D extends IDatabaseDriver = IDatabaseDriver> {
 
   private readonly validator = new EntityValidator(this.config.get('strict'));
   private readonly repositoryMap: Record<string, EntityRepository<AnyEntity>> = {};
-  private readonly entityLoader = new EntityLoader(this);
+  private readonly entityLoader: EntityLoader = new EntityLoader(this);
```
Comment (B4nan) [inline lib/EntityManager.ts]: why is this needed?

File: lib/EntityManager.ts [hunk 7]
```diff
@@ -5,28 +5,29 @@ import { AbstractSqlDriver, IDatabaseDriver } from './drivers';
 import { EntityData, EntityMetadata, EntityName, AnyEntity, IPrimaryKey, FilterQuery, Primary, Dictionary } from './types';
 import { QueryBuilder, QueryOrderMap, SmartQueryHelper } from './query';
 import { MetadataStorage } from './metadata';
-import { Connection, Transaction } from './connections';
+import { Transaction } from './connections';
+import { MySqlDriver } from './drivers/MySqlDriver';
```
Comment (B4nan) [inline lib/EntityManager.ts]: forgotten unused import

File: lib/EntityManager.ts (hunk 1 above)
Comment (B4nan) [reply lib/EntityManager.ts]: can we cast it to `ReturnType<D['getConnection']>` instead?

File: lib/EntityManager.ts [hunk 8]
```diff
@@ -48,7 +49,13 @@ export class EntityManager {
 
   createQueryBuilder(entityName: EntityName<AnyEntity>, alias?: string, type?: 'read' | 'write'): QueryBuilder {
     entityName = Utils.className(entityName);
-    return new QueryBuilder(entityName, this.metadata, this.driver as AbstractSqlDriver, this.transactionContext, alias, type);
+
+    const driver = this.driver as IDatabaseDriver<any>;
```
Comment (B4nan) [inline lib/EntityManager.ts]: move this line above (so the blank line is before the `if` block). 

also why `IDatabaseDriver<any>`? if its just about TS complaining about the instanceof, you can cast to unknown or object.

File: lib/MikroORM.ts [hunk 9]
```diff
@@ -7,22 +7,22 @@ import { Configuration, Logger, Options } from './utils';
...
-    const orm = new MikroORM(options);
+  static async init<D extends IDatabaseDriver = IDatabaseDriver>(options: Options<D> | Configuration<D>): Promise<MikroORM<D>> {
+    const orm = new MikroORM<D>(options);
     const driver = await orm.connect();
 
     try {
       const discovery = new MetadataDiscovery(MetadataStorage.init(), orm.driver.getPlatform(), orm.config, orm.logger);
       orm.metadata = await discovery.discover();
-      orm.em = new EntityManager(orm.config, driver, orm.metadata);
+      orm.em = new EntityManager<D>(orm.config, driver, orm.metadata);
```
Comment (B4nan) [inline lib/MikroORM.ts]: here the `<D>` would be inferred from `driver` variable type, right?

File: lib/MikroORM.ts [hunk 10]
```diff
@@ -66,11 +66,21 @@ export class MikroORM {
   }
 
   getSchemaGenerator(): SchemaGenerator {
-    return new SchemaGenerator(this.driver as AbstractSqlDriver, this.metadata);
+    const driver = this.driver as IDatabaseDriver<any>;
```
Comment (B4nan) [inline lib/MikroORM.ts]: again, please maintain blank line before block statements. 

but here I would rather drop `driver` variable and do this instead:

    if (!(this.driver as unknown instanceof AbstractSqlDriver)) {
      throw new Error('Not supported by given driver');
    }

    return new SchemaGenerator(this.driver, this.metadata);

File: lib/MikroORM.ts [hunk 11]
```diff
@@ -66,11 +66,21 @@ export class MikroORM {
...
+    if (!(driver instanceof AbstractSqlDriver)) {
+      throw new Error('Not supported by given driver');
+    }
+
+    return new SchemaGenerator(driver, this.metadata);
   }
 
   getEntityGenerator(): EntityGenerator {
-    return new EntityGenerator(this.driver as AbstractSqlDriver, this.config);
+    const driver = this.driver as IDatabaseDriver<any>;
```
Comment (B4nan) [inline lib/MikroORM.ts]: same here

File: tests/EntityGenerator.test.ts [hunk 12]
```diff
@@ -40,4 +40,12 @@ describe('EntityGenerator', () => {
     await orm.close(true);
   });
 
+  test('not supported [mongodb]', async () => {
+    const orm = await initORMMongo();
+
```
Comment (B4nan) [inline tests/EntityGenerator.test.ts]: i would remove those blank lines here

File: tests/EntityGenerator.test.ts [hunk 13]
```diff
@@ -40,4 +40,12 @@ describe('EntityGenerator', () => {
...
+    expect(() => orm.getEntityGenerator()).toThrowError('Not supported by given driver');
+
```
Comment (B4nan) [inline tests/EntityGenerator.test.ts]: and here

File: tests/QueryBuilder.test.ts [hunk 14]
```diff
@@ -1015,6 +1015,14 @@ describe('QueryBuilder', () => {
     expect(qb.getAliasForEntity(Author2.name, node)).toBeUndefined();
   });
 
+  test('not supported [mongodb]', async () => {
+    const orm = await initORMMongo();
+
```
Comment (B4nan) [inline tests/QueryBuilder.test.ts]: again, please remove those blank lines here

File: tests/QueryBuilder.test.ts [hunk 15]
```diff
@@ -1015,6 +1015,14 @@ describe('QueryBuilder', () => {
...
+    expect(() => orm.em.createQueryBuilder(Author2, 'a')).toThrowError('Not supported by given driver');
+
```
Comment (B4nan) [inline tests/QueryBuilder.test.ts]: and here

File: tests/SchemaGenerator.test.ts [hunk 16]
```diff
@@ -255,4 +255,12 @@ describe('SchemaGenerator', () => {
     await orm.close(true);
   });
 
+  test('not supported [mongodb]', async () => {
+    const orm = await initORMMongo();
+
```
Comment (B4nan) [inline tests/SchemaGenerator.test.ts]: and again here

File: tests/SchemaGenerator.test.ts [hunk 17]
```diff
@@ -255,4 +255,12 @@ describe('SchemaGenerator', () => {
...
+    expect(() => orm.getSchemaGenerator()).toThrowError('Not supported by given driver');
+
```
Comment (B4nan) [inline tests/SchemaGenerator.test.ts]: and here

File: lib/EntityManager.ts (hunk 6 above)
Comment (pepakriz) [reply lib/EntityManager.ts]: `error TS7022: 'entityLoader' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.`

File: lib/EntityManager.ts (hunk 8 above)
Comment (pepakriz) [reply lib/EntityManager.ts]: Ok, I've replaced it by `object` type

File: lib/MikroORM.ts (hunk 9 above)
Comment (pepakriz) [reply lib/MikroORM.ts]: right

File: lib/EntityManager.ts (hunk 1 above)
Comment (pepakriz) [reply lib/EntityManager.ts]: Wow, it works! Truly magic! :smiley:

File: tests/SchemaGenerator.test.ts (hunk 17 above)
Comment (pepakriz) [reply tests/SchemaGenerator.test.ts]: I've preserved this new line to be consistent with other tests.

File: tests/QueryBuilder.test.ts (hunk 15 above)
Comment (pepakriz) [reply tests/QueryBuilder.test.ts]: I've preserved this new line to be consistent with other tests.
